File: src/backend/Lonely.Api/Moderation/ModerationService.cs
namespace Lonely.Api.Moderation;

public record SubmitReportRequest(string ReporterId, string TargetId, string Reason, string? ContentId);

public record ReportResponse(
    string ReportId,
    string ReporterId,
    string TargetId,
    string Reason,
    string? ContentId,
    string Status);

public interface IModerationService
{
    Task<ReportResponse> SubmitReport(SubmitReportRequest request);
    Task<IEnumerable<ReportResponse>> GetQueue();
    Task<ReportResponse> Resolve(string reportId, string resolution);
    Task<IEnumerable<ReportResponse>> GetHistory(string userId);
}

public class ModerationService : IModerationService
{
    private readonly Dictionary<string, ReportResponse> _reports = new();

    public Task<ReportResponse> SubmitReport(SubmitReportRequest request)
    {
        var report = new ReportResponse(
            Guid.NewGuid().ToString(),
            request.ReporterId,
            request.TargetId,
            request.Reason,
            request.ContentId,
            "pending");
        _reports[report.ReportId] = report;
        return Task.FromResult(report);
    }

    public Task<IEnumerable<ReportResponse>> GetQueue() =>
        Task.FromResult(_reports.Values.Where(r => r.Status == "pending"));

    public Task<ReportResponse> Resolve(string reportId, string resolution)
    {
        if (!_reports.TryGetValue(reportId, out var report))
            throw new KeyNotFoundException($"Report {reportId} not found.");
        var updated = report with { Status = resolution };
        _reports[reportId] = updated;
        return Task.FromResult(updated);
    }

    public Task<IEnumerable<ReportResponse>> GetHistory(string userId) =>
        Task.FromResult(_reports.Values.Where(r => r.ReporterId == userId));
}

File: src/mobile/lib/features/moderation/moderation_service.dart
class ModerationReport {
  final String reportId;
  final String reporterId;
  final String targetId;
  final String reason;
  final String status;

  const ModerationReport({
    required this.reportId,
    required this.reporterId,
    required this.targetId,
    required this.reason,
    required this.status,
  });
}

abstract class ModerationService {
  Future<ModerationReport> submitReport(
      String reporterId, String targetId, String reason);
  Future<List<ModerationReport>> getQueue();
  Future<List<ModerationReport>> getHistory(String userId);
  Future<ModerationReport> resolve(String reportId, String resolution);
}
